lightning/common/pause.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// The implementation is based on https://github.com/golang/sync/blob/master/semaphore/semaphore.go
```

Comment:
maybe you can use https://godoc.org/sync#Cond

Reply:
I've tried Cond but it is much less efficient than the current solution (which is why I've included two benchmarks)

Current solution:

```
PASS: pause_test.go:144: pauseSuite.BenchmarkWaitCtxCanceled	10000000	       254 ns/op
PASS: pause_test.go:136: pauseSuite.BenchmarkWaitNoOp	500000000	         3.09 ns/op
```

The issue is that `(*sync.Cond).Wait()` doesn't take a context as input, so you'll need to create another goroutine and wait for a channel (see `configlist.go` for how we used `sync.Cond`). So the code would still be quite complicated.